README.txt
VK Audios Downloader
This python script download all audios from your account to specified in config folder.

Notice: Old version suddenly stoped working, so been rewritten from null at 04.08.2016 (now it absolutely wroten by me software, not fork)

Software requirements:
Python 2.7 with additional modules: vk (install with: "pip install vk")

P.S. This program stable working on my home server, that running Ubuntu 16.04.

Warning! Before executing saver.py: edit config.py and set your account login information and download destination dir.


CHANGELOG.md
VK Audios Saver python script changelog:

1 Nov 2016:

* Fixed 'Too many requests per second' VK API error while processing album playlists
* Added deleted audios lyrics autodeletion


15 Sep 2016:

* Added by VK Audio Albums playlists generation option
* Added audios lyrics saving option
* Added excess files autodeletion option

1 Sep 2016:

+ Changed autogen playlists extension from "m3u" to "m3u8" for better readability by most players (8 digit at the end gives them to know about file encoded with utf-8)

24 Aug 2016:

+ Changed mp3 download lib from "urlgrabber" to "urllib"

23 Aug 2016:

+ Changed VK API interraction lib from "vk_api" to "vk"

20 Aug 2016:

+ Fixed excess audios search (since 04 Aug 2016 update if song already downloaded it wont be marked as existing in VK audios)
+ Added solution for "VK hides near half of audios from response"
	Problem details:
		For correct audios response you should use any offical apps APP ID, that allows to get token.
		In other case VK servers won't give all your audios in resoponse of this script request.
		Real experience at 20 Aug 2016: from my 822 audios BEFORE it returned for download 498-500 audios only, but AFTER it return 818 (where it lost last 4 I still cannot understand).
	Solution details:
		Default App ID at access_token request link changed to Windows App's App ID.
  Special thanks to Elisey Izumtsev for this solution.
+ Added download error handler: in case of VK server don't give download link (it could occur, for example, if song is seized by administration to comply with the DMCA)

07 Aug 2016:

+ Security improvement: Auth moved from username&password in config to access_token usage
+ Now script not turn page to online state when start (problem has existed since 04 Aug 2016 update)

04 Aug 2016:

+ Project renamed from "VKAudioSync" to "VK Audios Saver python script"
+ Repository renamed from "vkaudiosync" to "vk-audios-saver-cli"
+ Script absolutely rewritten from null, now using vk_api lib for interraction with VK.com

18 Mar 2016:

+ Modified output messages locale
+ Added excess files & audios search in music download folder to runtime stat output

01 Mar 2016:

+ Added showing download status (current download percentage, current song number, total songs number)
+ Rewritten download errors handlers for optimization purposes
+ Added download error hander: if save song with default file format impossible, save with template "<AudioID>.mp3"


28 Feb 2016:

+ Rewritten audios "Artist" and "Title" values reading
+ Added multi-spaces remove for "Artist" and "Title" values ("Stuff   a" with 2 or more spaces converts to "Stuff a" with one space)

22 Feb 2016:

+ Removed unix commands usage for cross-platfrom run ability.
+ Output locale changed from English to Russian.
+ Output filename format changed from <AudioID>.mp3 to "<Artist> - <Title>.mp3" where it possible
+ Implemented safe audios "Artist" and "Title" end values convertation (remove unsafe symbols for end file path).


31 Jan 2016:

+ Repository creation
+ Publication of my "VKAudioSaver" script sources in GitHub repo


sync.py
# -*- coding: utf-8 -*-
import os
import urllib2
import json
import codecs
import re
import HTMLParser
from mutagen.easyid3 import EasyID3
from mutagen.id3 import ID3NoHeaderError
import urlgrabber
import sys
reload(sys)  
sys.setdefaultencoding('utf8')
import vk_api
import config

from time import time

def get_token(credentials=[]):
	if not len(credentials):
		credentials = ('friends',)

	token, user_id = vk_api.auth(config.USERNAME,
								 config.PASSWORD,
								 config.CLIENT_ID,
								 ",".join(credentials))	
	return (token, user_id)


def get_audio(token, uid):
	res = vk_api.call_method('audio.getCount', {
		'oid': uid}, token)
	audio_cnt = res['response']

	res = vk_api.call_method('audio.get', {
		'count': audio_cnt}, token)

	return res

print "Типо самопальная качалка музончега с ВКонтактика by MelnikovSM"
if not os.path.exists(config.MUSIC_PATH):
	os.makedirs(config.MUSIC_PATH)
print "Гружу список аудиозаписей со страницы.."
token, user_id = get_token(['audio'])
user_audio = get_audio(token, user_id)['response']
print "Качаю аудиозаписи в заданный в конфиге каталог: %s" % (config.MUSIC_PATH)
i = 0
ei = 0
afi = 0

flist=[f for f in os.listdir(config.MUSIC_PATH) if os.path.isfile(os.path.join(config.MUSIC_PATH, f))]

def diff(a, b):
	b = set(b)
	return [aa for aa in a if aa not in b]

rlist=[]

fp=codecs.open(config.MUSIC_PATH+"/"+config.PLAYLIST, 'w', 'utf-8')
fp.write("#EXTM3U\n")
for track in user_audio:
	i+=1
	aid = str(track.get('aid'))
	artist = re.sub(' +', ' ', (str(track.get('artist').encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
	title = re.sub(' +', ' ', (str(track.get('title').encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')

	if (artist=="" or artist.isspace()) and (title=="" or title.isspace()):
		fname=aid
	elif (artist=="" or artist.isspace()) and not (title=="" or title.isspace()):
		fname = re.sub(' +', ' ', ("Unknown - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))
	else: fname = re.sub(' +', ' ', (artist+" - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))
	
	url = track.get('url')
	filename = os.path.basename(url).split('?')[0]
	filepath = os.path.join(config.MUSIC_PATH, "%s.mp3" % (fname))
	print "["+str(int((i*1.0/len(user_audio)*1.0)*100))+"%] "+"Качаю песню \"%s\".. (#%s из %s)" % (fname, i, len(user_audio))
	g = urlgrabber.grabber.URLGrabber(reget='simple')
	nic=1
	if os.path.isfile(filepath)==False:
		try:
			g.urlgrab(url, filename=filepath)
		except urlgrabber.grabber.URLGrabError, e:
			try:
				if e.exception[1] != 'The requested URL returned error: 416 Requested Range Not Satisfiable':
					print('Ошибка закачки: '+e.exception[1])
					nic=0
					ei+=1
				else:
					print "Данная песня уже закачана, ничо не делаю.."
					afi+=1
			except AttributeError:
				nic=2
				print "Какая-то хрень с именем файла. Пофиг, сохраню как "+aid+".mp3"
				filepath = os.path.join(config.MUSIC_PATH, "%s.mp3" % (aid))
				g = urlgrabber.grabber.URLGrabber(reget='simple')
				if os.path.isfile(filepath)==False:
					try:
						g.urlgrab(url, filename=filepath)
					except urlgrabber.grabber.URLGrabError, e:
						if e.exception[1] != 'The requested URL returned error: 416 Requested Range Not Satisfiable':
							print('Ошибка закачки: '+e.exception[1])
							ei+=1
						else:
							print "Данная песня уже закачана, ничо не делаю.."
							afi+=1
					else:
						print "Данная песня уже закачана, ничо не делаю.."
						afi+=1
			pass
	else:
		print "Данная песня уже закачана, ничо не делаю.."
		afi+=1
	if nic==1:
		fp.write("#EXTINF:,%s\n" % (artist+" - "+title))
		fp.write("%s.mp3\n" % (fname))
		rlist.append("%s.mp3" % (fname))
	elif nic==2: 
		fp.write("#EXTINF:,%s\n" % (artist+" - "+title))
		fp.write("%s.mp3\n" % (aid))
		rlist.append("%s.mp3" % (aid))
fp.close()
rlist.append(config.PLAYLIST)
print """>>>>> Закачка завершена! <<<<<
==========> Статистика работы <=========="""
print "Всего аудиозаписей на странице: %s" %(i)
if afi>0:
	print "Ранее скачаных аудиозаписей в каталоге: %s" %(afi)
print "Всего скачано/докачано: %s" %(i-afi-ei)
if ei>0:
	print "Ошибок закачки: %s" %(ei)

diffr=diff(flist, rlist)
if len(diffr)>0:
	diffm=[]
	difff=[]
	for name in diffr:
		if ".mp3" in name:
			diffm.append(name)
		else:
			difff.append(name)
	if len(diffm)>0:
		print "Найдено %s (возможно) лишних аудиозаписей:" %(len(diffm))
		for name in diffm: print name
		print "Возможно они были удалены одминами ВКонтакта по требованию копирастов, или вы просто удалили эти аудиозаписи со страницы. Что с ними делать, решать вам."
	if len(difff)>0:
		print "Найдено %s (возможно) лишних файлов:" % len(difff)
		for name in difff: print name


saver.py
# -*- coding: utf-8 -*-
print('VK Audios Saver Script by MelnikovSM')
import sys
reload(sys)  
sys.setdefaultencoding('utf8')
sys.stdout.write('Loading libs.. ')
import vk
import re, os, codecs
import urllib
from time import sleep
print('Done')

# Config file load
sys.stdout.write('Loading config file.. ')
try:
	import config
	print('Done')
except ImportError as error_msg:
	print('Error')
	raise ImportError, 'Please check config.py in current runtime dir!'

# Authorization & API capture
sys.stdout.write('Proceeding authorization at VK.. ')
autherr=vk.exceptions.VkAPIError
try:
	vk=vk.API(vk.Session(access_token=config.token), v='5.53', lang='ru', timeout=10)
	vk.audio.get() # test
	print('Done')
except autherr:
	print('Error')
	raise Exception, 'VK access_token auth error!'

# Audios load
sys.stdout.write('Fetching audios list from VK.. ')
totalAudiosCount=vk.audio.getCount(owner_id=vk.users.get()[0]['id'])

audios=vk.audio.get()['items']

print('Done')
if totalAudiosCount<>len(audios): print('Notice: VK servers provided information about '+str(len(audios))+' audios only of '+str(totalAudiosCount)+' total.')

if not os.path.exists(config.destdir):
	print('Notice: Download destination dir is not exists, attempting to create..')
	os.makedirs(config.destdir)

if config.writeSongsLyrics==True:
	if config.lyricsDirIsPathAbsolute==True: lyricsDir=config.lyricsDir+'/'
	else: lyricsDir=config.destdir+'/'+config.lyricsDir+'/'
	if not os.path.exists(lyricsDir):
		print('Notice: Lyrics dir is not exists, attempting to create..')
		os.makedirs(lyricsDir)

print('Starting '+str(len(audios))+' audios download..')

flist=[f for f in os.listdir(config.destdir) if os.path.isfile(os.path.join(config.destdir, f))]
fllist=[f for f in os.listdir(lyricsDir) if os.path.isfile(os.path.join(lyricsDir, f))]

def diff(a, b):
	b = set(b)
	return [aa for aa in a if aa not in b]

rlist=[]
rllist=[]
fp=codecs.open(config.destdir+"/"+config.playlist_prefix+'.m3u8', 'w', 'utf-8')
fp.write("#EXTM3U\n")

aeac=0
rtec=0
sadc=0
dnac=0

for audio in range(len(audios)):
	# converting artist & title to proper runtime
	artist = re.sub(' +', ' ', (str(audios[audio]['artist'].encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
	title = re.sub(' +', ' ', (str(audios[audio]['title'].encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
	if (artist=="" or artist.isspace()) and (title=="" or title.isspace()):
		fname=aid
	elif (artist=="" or artist.isspace()) and not (title=="" or title.isspace()):
		fname = re.sub(' +', ' ', ("Unknown - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))
	else: fname = re.sub(' +', ' ', (artist+" - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))

	sys.stdout.write("["+str(int((audio*1.0/len(audios)*1.0)*100))+"%] "+"Processing \"%s\" (#%s of %s).. " % (fname, audio, len(audios)))
	filepath = os.path.join(config.destdir, "%s.mp3" % (fname))
	fp.write("#EXTINF:,%s\n" % (artist+" - "+title))
	fp.write("%s.mp3\n" % (fname))
	if os.path.isfile(filepath)==False:
		try:
			try:
				urllib.urlretrieve(str(audios[audio]['url']), filepath)
				rlist.append("%s.mp3" % (fname))
				sadc+=1
				print('Complete')
			except IndexError as e:
				print('Download unavailable')
				dnac+=1
		except:
			rtec+=1
			print('Error')
	else:
		aeac+=1
		print('Already exists.')
		rlist.append("%s.mp3" % (fname))
	if config.writeSongsLyrics==True:
		lid=0
		try: lid=audios[audio]['lyrics_id']
		except KeyError: pass
		if lid<>0:
			if not os.path.exists(lyricsDir+fname+'.txt'):
				lf=codecs.open(lyricsDir+fname+'.txt', 'w', 'utf-8')
				lf.write(vk.audio.getLyrics(lyrics_id=lid)['text'].encode('utf8'))
				lf.close()
				rllist.append("%s.txt" % (fname))
				sleep(0.32)
			rllist.append("%s.txt" % (fname))

if config.writeSongsLyrics==True:
	for f in diff(fllist, rllist):
		os.remove(lyricsDir+f)

fp.close()
rlist.append(config.playlist_prefix+'.m3u8')
rlist.append(config.playlist_prefix+'_deprecated.m3u8')
print('Download complete!')

if config.generateAlbumPlaylists==True:
	print('Generating playlists from your VK Albums..')
	albums=vk.audio.getAlbums()
	if albums['count']==len(albums['items']): print('Total '+str(albums['count'])+' albums found')
	else: print('Notice: Only '+str(len(albums['items']))+' albums accessable, but server returned count value '+str(albums['count']))
	for album in albums['items']:
		atitle=re.sub(' +', ' ', (str(album['title'].encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
		print('Processing album "'+atitle+'"..')
		plf=codecs.open(config.destdir+"/"+config.playlist_prefix+'_'+atitle+'.m3u8', 'w', 'utf-8')
		rlist.append(config.playlist_prefix+'_'+atitle+'.m3u8')
		plf.write("#EXTM3U\n")
		try: calbum=vk.audio.get(album_id=album['id'])['items']
		except autherr as e:
			sleep(0.33)
			calbum=vk.audio.get(album_id=album['id'])['items']
		for audio in calbum:

			artist = re.sub(' +', ' ', (str(audio['artist'].encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
			title = re.sub(' +', ' ', (str(audio['title'].encode('utf8')).strip())).replace('&amp', '&').replace('&;', '&')
			if (artist=="" or artist.isspace()) and (title=="" or title.isspace()):
				fname=aid
			elif (artist=="" or artist.isspace()) and not (title=="" or title.isspace()):
				fname = re.sub(' +', ' ', ("Unknown - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))
			else: fname = re.sub(' +', ' ', (artist+" - "+title).translate(None, ':*?!@%$<>|+\\\"').replace('/', '-'))

			filepath = os.path.join(config.destdir, "%s.mp3" % (fname))
			if os.path.isfile(filepath)==True:
				plf.write("#EXTINF:,%s\n" % (artist+" - "+title))
				plf.write("%s.mp3\n" % (fname))
			else: print('Warning: "'+fname+'" was not found at download dest dir, skipping from adding')
print('Complete')
print('Displaying runtime stat..')
print('Total '+str(totalAudiosCount)+' audios detected at target ID')
if totalAudiosCount<>len(audios): print('But only '+str(len(audios))+' audios available for script.')
if dnac>0: print('And VK server not gived download URL for '+str(dnac)+' audios.')
if sadc>0: print('Total successful downloads: '+str(sadc))
if aeac>0: print('Total already downloaded audios: '+str(aeac))
if rtec>0: print('Total terminated by error downloads: '+str(rtec))

diffr=diff(flist, rlist)

if len(diffr)>0:
	diffm=[]
	difff=[]
	for name in diffr:
		if ".mp3" in name:
			diffm.append(name)
		else:
			difff.append(name)
	if len(diffm)>0:
		print "Notice: Script found %s (maybe) excess audio files:" %(len(diffm))
		if config.deleteExcess==False:
			fp=codecs.open(config.destdir+"/"+config.playlist_prefix+"_deprecated.m3u8", 'w', 'utf-8')
			fp.write("#EXTM3U\n")
		for name in diffm:
				print name
				if config.deleteExcess==False: fp.write(name+"\n")
				else:
					os.remove(config.destdir+'/'+name)
		if config.deleteExcess==False: fp.close()
		if config.deleteExcess==False: print "It may be deleted or seized by VK administration audios. Saved to playlist deprecated."
		else:
			print "Excess deletion is enabled in config, it all were wiped"

	if len(difff)>0:
		print "Notice: Script found %s (maybe) excess files:" % len(difff)
		for name in difff:
			print name
			if config.deleteExcess==True: os.remove(config.destdir+'/'+name)
		if config.deleteExcess==True: print "Excess deletion is enabled in config, it all were wiped"


config_sample.py
# -*- coding: utf-8 -*-

# VK Audios Saver configuration file

# How to provide access to your audios to VK Audios Saver script:

# Go to this link then grant access:
# https://oauth.vk.com/authorize?client_id=3697615&scope=audio,offline&response_type=token
# Then copy access_token value from address link and put into token value below
# P.S. Read this instruction for more correct understanding: http://zenno.pro/kak-poluchit-access-token-prilozheniya-vk-com/

# Notice: You should use any offical apps APP ID, that allows to get token (Windows 8-10 PC/Mobile app ID is set by default, but servers don't allow to get token from Android APP ID, for example).
# In other case VK servers won't give all your audios in resoponse of this script request.
# Real experience at 20 Aug 2016: from my 822 audios BEFORE it returned for download 498-500 audios only, but AFTER it return 818 (where it lost last 4 I still cannot understand).
# This solution offered by Elisey Izumtsev. Huge human thanks to him.

# VK Access Information

# access_token value here
token='<insert your access_token value here>'


# Download config

# audios download destination dir
destdir='music'

# delete excess audios and files yes - True, no - False
deleteExcess=False

# playlist prefix
playlist_prefix='playlist'

# generate playlists from your VK Albums? yes - True, no - False
generateAlbumPlaylists=True


# write songs lyrics (if present) to specified dir with song name
writeSongsLyrics=True

# lyrics dir
lyricsDir='lyrics'
# Is lyrics dir location is absolute and should'nt be located inside download destination dir? yes - True, no - False
lyricsDirIsPathAbsolute=False